### DISCUSSION TITLE
Tying to use select and include together to create a partial structure

### DISCUSSION THREAD

**Discussion by Tanavya** [prisma/prisma#23931] [Q&A] [answered]
### Question

Following this guide https://www.prisma.io/docs/orm/prisma-client/type-safety/operating-against-partial-structures-of-model-types. I implement the following:

```
import { Prisma } from '@prisma/client'

const userArgs = Prisma.validator<Prisma.UserDefaultArgs>()({
  select: { email: true, name: true },
  include: { posts: true },
});

type UserInputData = Prisma.UserGetPayload<typeof userArgs>;
```

Basically I want a `UserInputData` type which can easily be passed into my create user function, which also creates other relations. But I don't want to have to pass in all fields, for instance id, createdAt, updatedAt, etc which can be initiated by default. 

However, for some reason the `UserInputData` type I create with the code above actually includes all the fields, as if specifying `select` had no effect on it. I think I could do some kind of hacky solution to combine the selected fields with the type generated with just include, but I was wondering if there is a better way to do what I want natively in Prisma. Is there a way to get this to work?

**Comment by Tanavya** [comment]
Okay, it seems like I got it to work. I needed to use _just_ select
```
const userArgs = Prisma.validator<Prisma.UserDefaultArgs>()({
  select: { email: true, name: true, posts: true },
});
```
However, now I have 2 further questions.
1. Why doesn't the original method work? Am I misunderstanding something?
2. How do I get a type where some fields are optional and can also include relations? For example to reiterate, for my use case I would like fields like `id`, `createdAt`, `updatedAt` to be optional. So I want a more general type like `UserModel` which includes relations and each sub-model and the "root" level `User` model can have fields that can be undefined, but are still required in the schema. This way I can use the same model as both input and output for create and get methods.

**Comment by ludralph** [comment, marked as answer]
Hi @Tanavya 👋 

> Why doesn't the original method work? Am I misunderstanding something?

The original method you attempted, which included both `select` and `include` operations, didn't produce the desired result because the `select` operation is primarily used to specify which scalar fields to include in the query result, while the `include` operation is used to specify related models or fields to include in the query result.

When you used both `select` and `include` together in the validator definition, the `include` operation might have overridden the `select` operation, resulting in all fields being included in the generated type, including those specified in the select operation.

In contrast, when you used only the `select` operation in the validator definition:
```ts
const userArgs = Prisma.validator<Prisma.UserDefaultArgs>()({
  select: { email: true, name: true, posts: true },
});
```
You explicitly specified the fields you wanted to include in the generated type, which ensured that only those fields were included and that the `include` operation didn't interfere with the selection of scalar fields.


> How do I get a type where some fields are optional and can also include relations? For example to reiterate, for my use case I would like fields like id, createdAt, updatedAt to be optional. So I want a more general type like UserModel which includes relations and each sub-model and the "root" level User model can have fields that can be undefined, but are still required in the schema. This way I can use the same model as both input and output for create and get methods.

If you want to create a type where some fields are optional and can also include relations, you would have to manually create these types in your application code. Here's an example:

```ts
type PartialUser = Partial<Prisma.UserCreateInput & {{
  posts: Prisma.PostCreateNestedManyWithoutAuthorInput;
}}>;
````
In this example, `PartialUser` is a type that includes all fields of the `User` model and the `posts` relation, and all these fields are optional. However, this approach increases the maintenance burden upon changes to the Prisma schema as you need to manually maintain the types.
